File: simple/simple-numerical.R
library(deSolve)
library(flexsurv) # For pgompertz

ss_death <- read.csv("ss-death-2011.csv")

inst_rate <- function(percent, timeframe)
{
  - log(1-percent) / timeframe
}

###################################
# Numerical approach to secular death (very high accuracy!)
f_40yr_percent_d    <- c(ss_death$f_death_prob[41:120])
sim_adj_age         <- 0:79 + 0.5 # 0.5 offset since percentage is for whole year
f_40yr_per_d_spline <- splinefun(sim_adj_age, f_40yr_percent_d)
plot(1:2); dev.off()
curve(f_40yr_per_d_spline, col='red', from=0, to=82, xlab="years past 40", ylab="percent chance of death")
points(sim_adj_age, f_40yr_percent_d)

# Clamped at infinite rate via pmin
f_40yr_drate <- function(t) inst_rate(pmin(f_40yr_per_d_spline(t), 1),1)
curve(f_40yr_drate, from=0, to=90)

####################################
# Integrations of death rates for exposure calculations in delay
# Now, a special function used in delay equation (Had to put upper bound at 81)
F_40yr_drate_5yr <- Vectorize(function(t)
{
  integrate(f_40yr_drate, lower=max(t-5, 0), upper=min(t, 81))$value
})


F_40yr_drate_1yr <- Vectorize(function(t)
{
  integrate(f_40yr_drate, lower=max(t-1, 0), upper=min(t, 81))$value
})


# Does a spline work faster?

x <- 0:160 / 2
y <- F_40yr_drate_5yr(x)
plot(x, y, typ='l')
f <- splinefun(x, y)
curve(f, add=TRUE, col='red', lty=2)
F_40yr_drate_5yr <- f


x <- 0:160 / 2
y <- F_40yr_drate_1yr(x)
plot(x, y, typ='l')
f <- splinefun(x, y)
curve(f, add=TRUE, col='red', lty=2)
F_40yr_drate_1yr <- f

# This is for doing numberical integration of a set of numbers at an even interval
alt_simp_coef <- function(i)
{
  if (i < 8) stop("Invalid Simpson coefficient size")
  
  # pg.117 4.1.14, Numerical Recipes in C, 1st edition
  c(17/48, 50/48, 43/48, 49/48, rep(1, i-8), 49/48, 43/48, 50/48, 17/48) 
}

###################################
# Main model parameters
params <- c(
  r_a  = inst_rate(0.1, 5),  # Rate of healthy having event A
  r_b  = inst_rate(0.5, 5),  # Rate of post-A  having event B
  r_ad = 0.05,               # Rate of death as direct result of A
  
  c_a  = 10000,              # Cost of event A
  c_b  = 25000,              # Cost of event B for a year
  c_t  = 0,                  # Cost of therapy
  d_a  = 0.25,               # Permanent disutility for a
  d_b  = 0.1,                # 1-year disutility for b 
  
  disc_rate = 1e-12          # For computing discount
)

###################################
# Numerical Delay Differential Equation
Simple <- function(t, y, params)
{
  with(as.list(c(y, params)), {
    
    # Use table for death_prob, Female 40 (offset 1)
    r_d <- f_40yr_drate(t)
    if(is.infinite(r_d)) r_d <- 1e16 # A really large number

    # Event B stops at time 5 years after event A (delay equation)
    dd_b <- if (t < 5 || t> 10) 0 else (1-r_ad)*r_a*lagvalue(t-5, 2)*exp(-5*r_b - F_40yr_drate_5yr(t)) 
    
    # Event A stops at time t=5 years
    if(t > 5) r_a <- 0
      
    list(c(
            disc = -disc_rate*disc,            # Simple discount r ate
            h    = -(r_a+r_d)*h,
            a    = r_a*h,
            e10  = (1-r_ad)*r_a*h-r_d*e10-dd_b-r_b*e10,
            e15  = dd_b - r_d*e15,
            b    = r_b*e10,
            e2   = r_b*e10 - r_d*e2,
            d    = r_ad*r_a*h + r_d*(h+e10+e15+e2),
            db   = r_b*e10 - r_d*db - if (t < 1) 0 else lagderiv(t-1, 6)*exp(-F_40yr_drate_1yr(t))
          )
    )
  })
}

yinit <- c(disc=1, h=1, a=0, e10=0, e15=0, b=0, e2=0, d=0, db=0)
times <- seq(0, 80, by=1/365)  # units of years, increments of days, everyone dies after 120, so simulation is cut short
system.time(out <- dede(yinit, times, Simple, params)) #, control=list(mxhist=1e6)))
#out   <- dede(yinit, times, Simple, params, control=list(mxhist=1e6))

plot(out)

# Check sensibility, i.e. all occupancy buckets sum to 1
all((rowSums(out[,c('h','e10','e15', 'e2','d')]) - 1) < 1e-8)

stop("Working halt")

costs <- function(solution, params)
{
  n <- length(solution[,1])
  simpson <- alt_simp_coef(n)

  with(as.list(params), {
    # Compute Discounted Cost
    cost <- c_a*sum(diff(solution[,'a'])*solution[2:n,'disc']) + # Cost * Number of events in bucket a
            c_b*sum(diff(solution[,'b'])*solution[2:n,'disc']) + # Cost * Number of events in bucket b
            c_t*solution[1, 'h']  # Therapy Cost * Initial healthy individuals
    
    # Step size of simulation
    step     <- solution[2,'time'] - solution[1,'time']
    
    # Total possible life units is integral of discounted time
    life <- sum(simpson*solution[,'disc'])*step
    
    # Permanent disutility for A (integration)
    disA <- d_a*sum(simpson*solution[,'e10']*solution[,'disc'])*step + 
            d_a*sum(simpson*solution[,'e15']*solution[,'disc'])*step +
            d_a*sum(simpson*solution[,'e2' ]*solution[,'disc'])*step
    
    # Event B
    disB <- d_b*sum(simpson*solution[,'db']*solution[,'disc'])*step

    # Death disutility
    disD <- sum(simpson*solution[,'d']*solution[,'disc'])*step       # Death disutility (integration)
    
    c(cost       = unname(cost),
      qaly       = unname(life-disA-disB-disD),
      possible   = unname(life),
      disutility = unname(disA+disB+disD),
      a_count    = unname(solution[n,'a']),
      disutil_a  = unname(disA),
      b_count    = unname(solution[n,'b']),
      disutil_b  = unname(disB),
      dead_count = unname(solution[n,'d']), 
      disutil_d  = unname(disD),
      living     = unname(solution[n,'h']+solution[n,'e10']+solution[n,'e15']+solution[n,'e2'])
      )
  })
}

expected <- function(params) costs(dede(yinit, times, Simple, params), params)

round(expected(params), 4)

params['r_a'] <- params['r_a']*0.5
params['c_t']  <- 2000
round(expected(params), 4)

# Some diagnostic plots
# 
# dev.off()
# plot(out[,'time'],out[,'db'], typ='l', xlim=c(0, 12))
# abline(h=0, col='red', lty=2)
# abline(v=11, col='blue', lty=2)
# abline(v=5, col='green', lty=2)
# 
# dev.off()
# plot(out[,'time'],out[,'b'], typ='l', xlim=c(0, 12))
# abline(v=10, col='green', lty=2)
# 
# dev.off()
# plot(out[,'time'],out[,'e10'], typ='l', xlim=c(0, 12))
# abline(v=10, col='green', lty=2)

File: simple/simple-numerical.jl
################################################
# Packages Required
# Pkg.add("DifferentialEquations")
# Pkg.add("DataFrames")
# Pkg.add("Dierckx")
# Pkg.add("Plots")

# From: Christopher Rackauckas
# If the problem was stiff, I think the best option right now is the algorithm
# alg = MethodOfSteps(Rosenbrock23()). Rosenbrock23() on ODEs isn't as fast as
# using LSODA, but you can't use LSODA in MethodOfSteps. 

using DataFrames
using DifferentialEquations
using Dierckx
using Plots

# Read 2011 SS death tables
ss_death = readtable("ss-death-2011.csv")

# Instantaneous exponential rate from percent over some time frame
instantaneous_rate(percent, timeframe) = - log(1-percent) / timeframe

################################################
# Parameters



#################################################
# Determine death rate function via spline
f_40yr_percent_d    = ss_death[:_f_death_prob][41:120]

sim_adj_age         = [0:79...] + 0.5 # 0.5 offset since percentage is for whole year
f_40yr_death_spline = Spline1D(sim_adj_age, f_40yr_percent_d)

# Test plot
# plot(sim_adj_age, f_40yr_death_spline(sim_adj_age))

f_40yr_death_rate(t) = instantaneous_rate(f_40yr_death_spline(t), 1)


####################################
# Integrations of death rates for exposure calculations in delay
# Now, a special function used in delay equation (Had to put upper bound at 81)
f_40yr_death_rate_5yr(t, r_b) = exp(-5.0*r_b-quadgk(f_40yr_death_rate, maximum([t-5, 0]), minimum([t, 81]))[1])
f_40yr_death_rate_1yr(t)      = exp(-quadgk(f_40yr_death_rate, maximum([t-1, 0]), minimum([t, 81]))[1])

####################################
# Define Model

function simple(t, y, h, p, dy)
  r_a  = p[1]
  r_b  = p[2]
  r_ad = p[3]
  r_d = f_40yr_death_rate(t)

  dd_b =  if (t >= 5.0 && t <= 10.0)
            (1-r_ad)*r_a*f_40yr_death_rate_5yr(t,r_b)*(h(t-5)[1])
          else
            0.0
          end

  # Local rate of a is function of time (it goes off at 5)
  lr_a = if(t > 5.0) 0.0 else r_a end

  dy[1] = -(lr_a+r_d)*y[1]                                # H bucket
  dy[2] = lr_a*y[1]                                       # A bucket
  dy[3] = (1-r_ad)*lr_a*y[1] - (r_b+r_d)*y[3] - dd_b      # E10 Bucket
  dy[4] = dd_b - r_d*y[4]                                 # E15 Bucket
  dy[5] = r_b*y[3] - r_d*y[5]                             # E20 Bucket
  dy[6] = r_b*y[3]                                        # B bucket
  dy[7] = r_b*y[3] - r_d*y[7]                             # Disutility of B
  
  if(y[3] < 0) 
    dy[3] = -y[3]
  end
  
  if(t>=1 && t <= 11) 
    dy[7] = dy[7] - (r_b*h(t-1)[3])*f_40yr_death_rate_1yr(t)
  end
end
            
pf_simple = DDEParameterizedFunction(
  simple,
  [instantaneous_rate(0.1, 5), instantaneous_rate(0.5, 5),  0.05])

hh(t) = zeros(7)

prob = ConstantLagDDEProblem(
  pf_simple,
  hh,
  vcat([1.0], zeros(6)),
  [1,5],
  (0.0, 80.0))
 
########## How to run 
sol=solve(prob, MethodOfSteps(Tsit5()))
plot(sol)

function discounted_a(sol, discount, a_rate)
  quadgk(function(t) a_rate*(sol(t)[1])*exp(-discount * t) end, 0, 5)[1]
end

function discounted_b(sol, discount, b_rate)
  quadgk(function(t) b_rate*(sol(t)[3])*exp(-discount * t) end, 0, 10)[1]
end

function disutility_a_time(sol, discount)
  quadgk(function(t) sol(t)[3]*exp(-discount*t) end, 0, 80)[1] +
  quadgk(function(t) sol(t)[4]*exp(-discount*t) end, 0, 80)[1] +
  quadgk(function(t) sol(t)[5]*exp(-discount*t) end, 0, 80)[1]
end

function disutility_b_time(sol, discount)
  quadgk(function(t) sol(t)[7]*exp(-discount*t) end, 0, 20)[1]
end

function healthy_time(sol, discount)
  quadgk(function(t) sol(t)[1]*exp(-discount*t) end, 0, 80)[1]
end

function outcome(sol, params)
  cost = params[:costT] + 
         params[:costA]*discounted_a(sol, params[:discount], params[:riskA]) +
         params[:costB]*discounted_b(sol, params[:discount], params[:riskB])
  
  qaly = healthy_time(sol, params[:discount]) +
         (1-params[:disA])*disutility_a_time(sol, params[:discount]) -
         params[:disB]*disutility_b_time(sol, params[:discount])

  Dict(
    :dCost => cost,
    :dQALY => qaly
  )
end

function params_as_dict(df)
  Dict(
    :riskA  => instantaneous_rate(df[1,:vRiskA],5),
    :riskB  => instantaneous_rate(df[1,:vRiskB],5),
    :fatalA => df[1,:vFatalA],
    :rrB    => df[1,:vRR],
    
    :costA  => 10000.0,
    :costB  => 25000.0,
    :costT  => 0.0,
    :disA   => 0.25,  # Permanent disutility for a
    :disB   => 0.1,   # 1-year disutility for b

    :discount => 0.03         # For computing discount
  )
end

function solution(df::DataFrames.DataFrame)

  params = params_as_dict(df)

  pf_simple.params[1] = params[:riskA]
  pf_simple.params[2] = params[:riskB]
  pf_simple.params[3] = params[:fatalA]
  
  prob = ConstantLagDDEProblem(
    pf_simple,
    hh,
    vcat([1.0], zeros(6)),
    [1,5],
    (0.0, 80.0))
    
  sol1=solve(prob, MethodOfSteps(Tsit5()))

  pf_simple.params[1] = pf_simple.params[1]*params[:rrB]
  
  sol2=solve(prob, MethodOfSteps(Tsit5()))
  
  (sol1, sol2)

  #[1,2,3]
end

#### Main Loop
cube = readtable("test-cube.csv")
for i in [1:(size(cube)[1])...]
  println(join([@sprintf "%.f" x for x in solution(cube[i,:])], ", "))
end

